vms/evm/warp/backend.go

```diff
@@ -0,0 +1,244 @@
+type Verifier struct {
+	db            *DB
+	blockClient   BlockStore
+	uptimeTracker *uptimetracker.UptimeTracker
+
+	// Metrics
+	messageParseFail            metrics.Counter
+	addressedCallValidationFail metrics.Counter
+	blockValidationFail         metrics.Counter
+	uptimeValidationFail        metrics.Counter
+}
+
+// NewVerifier creates a new warp message verifier.
+func NewVerifier(
+	db *DB,
+	blockClient BlockStore,
+	uptimeTracker *uptimetracker.UptimeTracker,
+) *Verifier {
+	return &Verifier{
+		db:                          db,
+		blockClient:                 blockClient,
+		uptimeTracker:               uptimeTracker,
+		messageParseFail:            metrics.NewRegisteredCounter("warp_backend_message_parse_fail", nil),
+		addressedCallValidationFail: metrics.NewRegisteredCounter("warp_backend_addressed_call_validation_fail", nil),
+		blockValidationFail:         metrics.NewRegisteredCounter("warp_backend_block_validation_fail", nil),
+		uptimeValidationFail:        metrics.NewRegisteredCounter("warp_backend_uptime_validation_fail", nil),
+	}
+}
+
+// Verify implements acp118.Verifier and validates whether a warp message should be signed.
+func (v *Verifier) Verify(ctx context.Context, unsignedMessage *warp.UnsignedMessage, _ []byte) *common.AppError {
+	messageID := unsignedMessage.ID()
+	// Known on-chain messages should be signed
+	if _, err := v.db.Get(messageID); err == nil {
+		return nil
+	} else if err != database.ErrNotFound {
+		return &common.AppError{
+			Code:    ParseErrCode,
+			Message: fmt.Sprintf("failed to get message %s: %s", messageID, err.Error()),
+		}
+	}
+
+	parsed, err := payload.Parse(unsignedMessage.Payload)
+	if err != nil {
+		v.messageParseFail.Inc(1)
+		return &common.AppError{
+			Code:    ParseErrCode,
+			Message: "failed to parse payload: " + err.Error(),
+		}
+	}
+
+	switch p := parsed.(type) {
+	case *payload.AddressedCall:
+		return v.verifyOffchainAddressedCall(p)
+	case *payload.Hash:
+		return v.verifyBlockMessage(ctx, p)
+	default:
+		v.messageParseFail.Inc(1)
+		return &common.AppError{
+			Code:    ParseErrCode,
+			Message: fmt.Sprintf("unknown payload type: %T", p),
+		}
+	}
+}
+
+// verifyBlockMessage returns nil if blockHashPayload contains the ID
+// of an accepted block indicating it should be signed by the VM.
+func (v *Verifier) verifyBlockMessage(ctx context.Context, blockHashPayload *payload.Hash) *common.AppError {
+	blockID := blockHashPayload.Hash
+	_, err := v.blockClient.GetBlock(ctx, blockID)
+	if err != nil {
+		v.blockValidationFail.Inc(1)
+		return &common.AppError{
+			Code:    VerifyErrCode,
+			Message: fmt.Sprintf("failed to get block %s: %s", blockID, err.Error()),
+		}
+	}
+
+	return nil
+}
+
+// verifyOffchainAddressedCall verifies the addressed call message
+func (v *Verifier) verifyOffchainAddressedCall(addressedCall *payload.AddressedCall) *common.AppError {
+	// Further, parse the payload to see if it is a known type.
```

Comment:
We should do a pass over this diff - but generally we should aim to document "why" code is doing things instead of "what" or "how" it does. "what/how" can be deduced from reading the code - and if it can't then it's likely the code should be written to be self-documenting. "why" is harder to reason about because it requires context that might not be immediately obvious so it is more valuable to document.